### DISCUSSION TITLE
Custom Phase

### DISCUSSION THREAD

**Discussion by felixdube** [pyuvm/pyuvm#203] [Q&A] [answered]
Hi,
I would like to add a phase to pyuvm. I saw the comment in s09_phasing.py.
```
# One could add phases by simply extending uvm_topdown_phase
# or uvm_bottom_up phase with a new phase named 'uvm_my_phase' and adding
# the my_phase() method to a uvm component with setattr.
```

Does it really need to be added with setattr? Could I just define my_phase() in one of my class that extend uvm_component? Also since uvm_my_phase is not in the uvm_common_phase list, will it really work? Does uvm_my_phase needs to be added to uvm_common_phase somehow?

I would also like this new phase to be an async method. My first idea was to have the uvm_my_phase extend either uvm_topdown_phase or uvm_bottom_up AND uvm_threaded_execute_phase, but I see that in run_test(), we are only awaiting for the uvm_run_phase. So I don't know if this is even possible to do.

**Comment by raysalemi** [comment]
> Hi,
> 
> I would like to add a phase to pyuvm. I saw the comment in s09_phasing.py.
> 
> ```
> 
> # One could add phases by simply extending uvm_topdown_phase
> 
> # or uvm_bottom_up phase with a new phase named 'uvm_my_phase' and adding
> 
> # the my_phase() method to a uvm component with setattr.
> 
> ```
> 
> 
> 
> Does it really need to be added with setattr? Could I just define my_phase() in one of my class that extend uvm_component? Also since uvm_my_phase is not in the uvm_common_phase list, will it really work? Does uvm_my_phase needs to be added to uvm_common_phase somehow?
> 
> 
> 
> I would also like this new phase to be an async method. My first idea was to have the uvm_my_phase extend either uvm_topdown_phase or uvm_bottom_up AND uvm_threaded_execute_phase, but I see that in run_test(), we are only awaiting for the uvm_run_phase. So I don't know if this is even possible to do.

Hi Felix,

In the end, anything is possible.  However, in 20 years I’ve never seen a solution that required additional phases.  Custom phases were defined for VIP developers, but were rarely used. 

I don’t understand your plan to extend both topdown/bottomup and threaded_execute.  UVM intended that you would extend only one. 

What are you designing?  What would you like to accomplish?

**Reply by felixdube** [reply]
Hi Ray,

Thank you for the quick reply!

I would like is to do some status register checks at the end of each tests. I usually do this in the check_phase using backdoor register access, but backdoor is not an option in the testbench I am currently working on. I also cannot use frontdoor access since the check_phase is not an async method.

I could explicitly call my own async register_check method at the end of the test's run_phase, but that would require the I modify all my tests, which I would like to avoid.

I am open to other suggestions.

Thanks,
Felix

**Comment by raysalemi** [comment]
Why not extend `uvm_test` to create `uvm_check_reg_test` and modify the `run_phase` there to call "super.run_phase()" and then await `async register_check`? 

Then you just change the first line of your tests with sed or some other batch editor?

**Reply by felixdube** [reply]
I am not sure I understand. Are you suggesting that my tests should then extend `uvm_reg_check_test` (they are currently extending `uvm_test`)? That would not work because the `run_phase` is overridden in the test classes. I think I really need to explicitly await `register_check` at the end of each test's `run_phase`, which is not obvious to do with sed or something like that.

Is this what you meant or I misunderstood?

**Reply by raysalemi** [reply, marked as answer]
Hi Felix,
Did you come up with a solution?  

Another solution is to change your `run_phase()` methods to some other name, then extend `uvm_test()` to create a `run_phase()` that awaits your async function and then calls the register checking function.

Ray

**Reply by felixdube** [reply]
Hi Ray,
Yes, that would work for my use case. I guess that’s the closest we can get to adding a phase without actually adding a phase.

Thank you for the help,
Felix

**Reply by raysalemi** [reply]
I think it is also a cleaner solution.  Messing with phasing would create a maintenance nightmare.
